Make inductor kernel metadata comments more descriptive

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126698
* #126573

Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code

Example float8-overhead-related inductor kernel comment after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/27/c27ts3rdw56ns7od5j6ovdnhxphished2lcu3adclzzixoo7khg5.py                                                                                      
# Source Nodes: [weight_fp8], Original ATen: [aten.mul, aten.clamp, aten._to_copy]                                                                                                        
# Source node to ATen node mapping:                                                                                                                                                       
#   weight_fp8 => clamp_max_1, clamp_min_3, convert_element_type_10, convert_element_type_11, convert_element_type_9, mul_3                                                               
# Graph fragment:                                                                                                                                                                         
#   %mul_3 : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%primals_2, %convert_element_type_8), kwargs = {})                                                   
#   %convert_element_type_9 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%mul_3, torch.float32), kwargs = {})                             
#   %clamp_min_3 : [num_users=1] = call_function[target=torch.ops.aten.clamp_min.default](args = (%convert_element_type_9, -448.0), kwargs = {})                                          
#   %clamp_max_1 : [num_users=1] = call_function[target=torch.ops.aten.clamp_max.default](args = (%clamp_min_3, 448.0), kwargs = {})                                                      
#   %convert_element_type_10 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%clamp_max_1, torch.bfloat16), kwargs = {})                     
#   %convert_element_type_11 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%convert_element_type_10, torch.float8_e4m3fn), kwargs = {})    
triton_poi_fused__to_copy_clamp_mul_5 = async_compile.triton('triton_', '''                                                                                                                                                                                                                                                                                                 
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

vkuzo committed:
Make inductor kernel metadata comments more descriptive

Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code
3. Adds a printout of the unique fwd and backward nn.Module stack
   elements, to make it easier to map the inductor kernel to the
   modeling code

Example inductor kernel comment from LLaMa 3 8B training after this PR:

```
triton_red_fused__to_copy_add_div_mean_mul_pow_rsqrt_sum_33 = async_compile.triton('triton_', '''
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
adding folks who recently touched this code as reviewers

vkuzo commented on torch/_inductor/utils.py:
```diff
@@ -256,11 +257,15 @@ def convert_shape_to_symint(
-        else int(i)
-        if isinstance(i, sympy.Integer)
-        else V.graph.sizevars.shape_env.create_symintnode(i, hint=None)
+        (
```
this is from linter, no logic change, can revert if needed

ezyang replied:
This file is UFMT enabled, why is the linter wobbling these lines 🤔

vkuzo commented on torch/_inductor/utils.py:
```diff
@@ -388,11 +393,9 @@ def sort_func(elem):
     @staticmethod
-    def clear_cache(self) -> None:
-        ...
+    def clear_cache(self) -> None: ...
```
from linter

vkuzo commented on torch/_inductor/utils.py:
```diff
@@ -388,11 +393,9 @@ def sort_func(elem):
 
-    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> RV:
-        ...
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> RV: ...
```
from linter

ezyang commented:
IDs in the code looks sus, 139645798654288, because it means the generated files are no longer deterministic

vkuzo commented:
> IDs in the code looks sus, 139645798654288, because it means the generated files are no longer deterministic

hmm, ok, let me see what we can do about that

Chillee commented:
We definitely want some tests for this btw.

ezyang commented on torch/_inductor/utils.py:
```diff
@@ -479,15 +489,56 @@ def get_kernel_metadata(node_schedule, wrapper):
+        node_s = node.format_node()
+        # Make the node string representation easier to read
+        # before: %abs_3 : [num_users=1] = call_function[target=torch.ops.aten.abs.default](args = (%tangents_1,), kwargs = {})
+        # after: %abs_3 = torch.ops.aten.abs.default(args = (%tangents_1,), kwargs = {})
```
It'd be nice to use the Python code formatter but I'm not sure if you can conveniently only do a few nodes without having to shove them in an fx.Graph or something

vkuzo replied:
yeah, I hacked it after looking at how much work it would be to implement it properly and with the context that this PR should not be in the critical path, but can take another look if you prefer, lmk

vkuzo replied:
made it a bit less hacky

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code
3. Adds a printout of the unique fwd and backward nn.Module stack
   elements, to make it easier to map the inductor kernel to the
   modeling code

Example inductor kernel comment from LLaMa 3 8B training after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/xy/cxyrctkcmf7tza4nypuaqokkfr5cz4pds7bds33zbv64fpdxad4s.py                                                         
# Source Nodes: [float_1, pow_1, mean, add, rsqrt], Original ATen: [aten._to_copy, aten.pow, aten.mean, aten.add, aten.rsqrt, aten.mul, aten.sum, aten.div]  
# Source node to ATen node mapping:                                                                                                                          
#   add => add                                                                                                                                               
#   float_1 => convert_element_type                                                                                                                          
#   mean => mean                                                                                                                                             
#   pow_1 => pow_1                                                                                                                                           
#   rsqrt => rsqrt                                                                                                                                           
# Graph fragment:                                                                                                                                            
#     %convert_element_type = torch.ops.prims.convert_element_type.default(args = (%primals_1, torch.float32), kwargs = {})                                  
#     %pow_1 = torch.ops.aten.pow.Tensor_Scalar(args = (%convert_element_type, 2), kwargs = {})                                                              
#     %mean = torch.ops.aten.mean.dim(args = (%pow_1, [-1, True), kwargs = {})                                                                               
#     %add = torch.ops.aten.add.Tensor(args = (%mean, 1e-05), kwargs = {})                                                                                   
#     %rsqrt = torch.ops.aten.rsqrt.default(args = (%add,), kwargs = {})                                                                                     
#     %add_8 = torch.ops.aten.add.Tensor(args = (%view_52, %view_54), kwargs = {})                                                                           
#     %add_9 = torch.ops.aten.add.Tensor(args = (%add_8, %view_56), kwargs = {})                                                                             
#     %mul_66 = torch.ops.aten.mul.Tensor(args = (%add_9, %primals_3), kwargs = {})                                                                          
#     %convert_element_type_121 = torch.ops.prims.convert_element_type.default(args = (%mul_66, torch.float32), kwargs = {})                                 
#     %mul_67 = torch.ops.aten.mul.Tensor(args = (%convert_element_type_121, %convert_element_type), kwargs = {})                                            
#     %mul_68 = torch.ops.aten.mul.Tensor(args = (%convert_element_type_121, %rsqrt), kwargs = {})                                                           
#     %sum_6 = torch.ops.aten.sum.dim_IntList(args = (%mul_67, [2, True), kwargs = {})                                                                       
#     %div_1 = torch.ops.aten.div.Scalar(args = (%expand_3, 4096), kwargs = {})                                                                              
#     %pow_6 = torch.ops.aten.pow.Tensor_Scalar(args = (%convert_element_type, 1.0), kwargs = {})                                                            
#     %mul_71 = torch.ops.aten.mul.Scalar(args = (%pow_6, 2.0), kwargs = {})                                                                                 
#     %mul_72 = torch.ops.aten.mul.Tensor(args = (%div_1, %mul_71), kwargs = {})                                                                             
#     %add_10 = torch.ops.aten.add.Tensor(args = (%mul_68, %mul_72), kwargs = {})                                                                            
#     %convert_element_type_122 = torch.ops.prims.convert_element_type.default(args = (%add_10, torch.bfloat16), kwargs = {})                                
#     %add_11 = torch.ops.aten.add.Tensor(args = (%add_7, %convert_element_type_122), kwargs = {})                                                           
# Unique node to nn_module_stack last element                                                                                                                
#     ('139645798654288', ("L['self']._checkpoint_wrapped_module.attention_norm", <class 'torchtitan.models.norms.RMSNorm'>))                                
# Unique node to fwd_nn_module_stack last element                                                                                                            
#     ('139645901077456', ("L['self']._checkpoint_wrapped_module.feed_forward.w2", <class 'float8_experimental.float8_dynamic_linear.Float8DynamicLinear'>)) 
#     ('139645798654288', ("L['self']._checkpoint_wrapped_module.attention_norm", <class 'torchtitan.models.norms.RMSNorm'>))                                
triton_red_fused__to_copy_add_div_mean_mul_pow_rsqrt_sum_33 = async_compile.triton('triton_', '''                                                            
import triton                                                                                                                                                
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo commented:
> IDs in the code looks sus, 139645798654288, because it means the generated files are no longer deterministic

hid behind a config, off by default

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code
3. Adds a printout of the unique fwd and backward nn.Module stack
   elements, to make it easier to map the inductor kernel to the
   modeling code

Example inductor kernel comment from LLaMa 3 8B training after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/xy/cxyrctkcmf7tza4nypuaqokkfr5cz4pds7bds33zbv64fpdxad4s.py                                                         
# Source Nodes: [float_1, pow_1, mean, add, rsqrt], Original ATen: [aten._to_copy, aten.pow, aten.mean, aten.add, aten.rsqrt, aten.mul, aten.sum, aten.div]  
# Source node to ATen node mapping:                                                                                                                          
#   add => add                                                                                                                                               
#   float_1 => convert_element_type                                                                                                                          
#   mean => mean                                                                                                                                             
#   pow_1 => pow_1                                                                                                                                           
#   rsqrt => rsqrt                                                                                                                                           
# Graph fragment:                                                                                                                                            
#     %convert_element_type = torch.ops.prims.convert_element_type.default(args = (%primals_1, torch.float32), kwargs = {})                                  
#     %pow_1 = torch.ops.aten.pow.Tensor_Scalar(args = (%convert_element_type, 2), kwargs = {})                                                              
#     %mean = torch.ops.aten.mean.dim(args = (%pow_1, [-1, True), kwargs = {})                                                                               
#     %add = torch.ops.aten.add.Tensor(args = (%mean, 1e-05), kwargs = {})                                                                                   
#     %rsqrt = torch.ops.aten.rsqrt.default(args = (%add,), kwargs = {})                                                                                     
#     %add_8 = torch.ops.aten.add.Tensor(args = (%view_52, %view_54), kwargs = {})                                                                           
#     %add_9 = torch.ops.aten.add.Tensor(args = (%add_8, %view_56), kwargs = {})                                                                             
#     %mul_66 = torch.ops.aten.mul.Tensor(args = (%add_9, %primals_3), kwargs = {})                                                                          
#     %convert_element_type_121 = torch.ops.prims.convert_element_type.default(args = (%mul_66, torch.float32), kwargs = {})                                 
#     %mul_67 = torch.ops.aten.mul.Tensor(args = (%convert_element_type_121, %convert_element_type), kwargs = {})                                            
#     %mul_68 = torch.ops.aten.mul.Tensor(args = (%convert_element_type_121, %rsqrt), kwargs = {})                                                           
#     %sum_6 = torch.ops.aten.sum.dim_IntList(args = (%mul_67, [2, True), kwargs = {})                                                                       
#     %div_1 = torch.ops.aten.div.Scalar(args = (%expand_3, 4096), kwargs = {})                                                                              
#     %pow_6 = torch.ops.aten.pow.Tensor_Scalar(args = (%convert_element_type, 1.0), kwargs = {})                                                            
#     %mul_71 = torch.ops.aten.mul.Scalar(args = (%pow_6, 2.0), kwargs = {})                                                                                 
#     %mul_72 = torch.ops.aten.mul.Tensor(args = (%div_1, %mul_71), kwargs = {})                                                                             
#     %add_10 = torch.ops.aten.add.Tensor(args = (%mul_68, %mul_72), kwargs = {})                                                                            
#     %convert_element_type_122 = torch.ops.prims.convert_element_type.default(args = (%add_10, torch.bfloat16), kwargs = {})                                
#     %add_11 = torch.ops.aten.add.Tensor(args = (%add_7, %convert_element_type_122), kwargs = {})                                                           
# Unique node to nn_module_stack last element                                                                                                                
#     ('139645798654288', ("L['self']._checkpoint_wrapped_module.attention_norm", <class 'torchtitan.models.norms.RMSNorm'>))                                
# Unique node to fwd_nn_module_stack last element                                                                                                            
#     ('139645901077456', ("L['self']._checkpoint_wrapped_module.feed_forward.w2", <class 'float8_experimental.float8_dynamic_linear.Float8DynamicLinear'>)) 
#     ('139645798654288', ("L['self']._checkpoint_wrapped_module.attention_norm", <class 'torchtitan.models.norms.RMSNorm'>))                                
triton_red_fused__to_copy_add_div_mean_mul_pow_rsqrt_sum_33 = async_compile.triton('triton_', '''                                                            
import triton                                                                                                                                                
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ezyang commented on torch/_inductor/utils.py:
```diff
@@ -471,6 +474,14 @@ def get_kernel_metadata(node_schedule, wrapper):
+    if len(inductor_nodes):
+        graph = inductor_nodes[0].graph
+        if all(n for n in inductor_nodes if n in graph.nodes):
+            inductor_nodes = [n for n in graph.nodes if n in inductor_nodes]
```
This is O(n) on the FX graph nodes, and so this is ultimately quadratic in the number of kernels in the Inductor graph

ezyang commented on torch/_inductor/utils.py:
```diff
@@ -479,15 +490,61 @@ def get_kernel_metadata(node_schedule, wrapper):
+        # A bit hacky: get the parent graph generated code and only keep the
+        # lines corresponding to the variables we care about. This allows us
+        # to reuse the Python codegen code.
+        for line in graph.python_code("").src.split("\n"):
```
Uhhh, this is a Graph, not a GraphModule, right? So the Python code is not cached, you're rendering it over and over again every time you print the aot autograd graph fragment. The Python code rendering is pretty expensive, definitely not recommended.

ezyang commented:
In the spirit of constructiveness, here is my suggestion for how to do the implementations without being heinously inefficient:

1. Just write a little node formatter by hand, it's not too hard. If you want to, you can factor out the python code formatter so that you can get at a helper for node by node
2. Generate a numbering for every node in the graph (node_1 => 0, node_2 => 1), cache it (I'm not sure where, you could potentially just stick it on the Graph if you're lazy), and then sort on these ints

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code
3. Adds a printout of the unique fwd and backward nn.Module stack
   elements, to make it easier to map the inductor kernel to the
   modeling code.  This is gated by `torch._inductor.config.comment_nn_module_stack_trace_info` and is off by default.

Example inductor kernel comment from LLaMa 3 8B training after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/km/ckm65ajvvhdoctx6djzhrkt7k5zmf3wyyugfqbhzyzplrd3eycxp.py                                                        
# Source Nodes: [silu, mul_6, abs_13, max_13], Original ATen: [aten.silu, aten.mul, aten.abs, aten.max]                                                     
# Source node to ATen node mapping:                                                                                                                         
#   abs_13 => abs_13                                                                                                                                        
#   max_13 => max_13                                                                                                                                        
#   mul_6 => mul_31                                                                                                                                         
#   silu => convert_element_type_58, convert_element_type_59, mul_26, sigmoid                                                                               
# Graph fragment:                                                                                                                                           
#     convert_element_type_58 = torch.ops.prims.convert_element_type.default(view_24, torch.float32)                                                        
#     sigmoid = torch.ops.aten.sigmoid.default(convert_element_type_58)                                                                                     
#     mul_26 = torch.ops.aten.mul.Tensor(convert_element_type_58, sigmoid);  convert_element_type_58 = sigmoid = None                                       
#     convert_element_type_59 = torch.ops.prims.convert_element_type.default(mul_26, torch.bfloat16);  mul_26 = None                                        
#     mul_31 = torch.ops.aten.mul.Tensor(convert_element_type_59, view_27);  convert_element_type_59 = view_27 = None                                       
#     abs_13 = torch.ops.aten.abs.default(mul_31)                                                                                                           
#     max_13 = torch.ops.aten.max.default(abs_13);  abs_13 = None                                                                                           
# Unique node to nn_module_stack last element                                                                                                               
#     ('139761509383760', ("L['self']._checkpoint_wrapped_module.feed_forward", <class 'torchtitan.models.llama.model.FeedForward'>))                       
#     ('139761460312400', ("L['self']._checkpoint_wrapped_module.feed_forward.w2", <class 'float8_experimental.float8_dynamic_linear.Float8DynamicLinear'>))
triton_red_fused_abs_max_mul_silu_23 = async_compile.triton('triton_', '''                                                                                                                                                                                                            
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code
3. Adds a printout of the unique fwd and backward nn.Module stack
   elements, to make it easier to map the inductor kernel to the
   modeling code.  This is gated by `torch._inductor.config.comment_nn_module_stack_trace_info` and is off by default.

Example inductor kernel comment from LLaMa 3 8B training after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/km/ckm65ajvvhdoctx6djzhrkt7k5zmf3wyyugfqbhzyzplrd3eycxp.py                                                        
# Source Nodes: [silu, mul_6, abs_13, max_13], Original ATen: [aten.silu, aten.mul, aten.abs, aten.max]                                                     
# Source node to ATen node mapping:                                                                                                                         
#   abs_13 => abs_13                                                                                                                                        
#   max_13 => max_13                                                                                                                                        
#   mul_6 => mul_31                                                                                                                                         
#   silu => convert_element_type_58, convert_element_type_59, mul_26, sigmoid                                                                               
# Graph fragment:                                                                                                                                           
#     convert_element_type_58 = torch.ops.prims.convert_element_type.default(view_24, torch.float32)                                                        
#     sigmoid = torch.ops.aten.sigmoid.default(convert_element_type_58)                                                                                     
#     mul_26 = torch.ops.aten.mul.Tensor(convert_element_type_58, sigmoid);  convert_element_type_58 = sigmoid = None                                       
#     convert_element_type_59 = torch.ops.prims.convert_element_type.default(mul_26, torch.bfloat16);  mul_26 = None                                        
#     mul_31 = torch.ops.aten.mul.Tensor(convert_element_type_59, view_27);  convert_element_type_59 = view_27 = None                                       
#     abs_13 = torch.ops.aten.abs.default(mul_31)                                                                                                           
#     max_13 = torch.ops.aten.max.default(abs_13);  abs_13 = None                                                                                           
# Unique node to nn_module_stack last element                                                                                                               
#     ('139761509383760', ("L['self']._checkpoint_wrapped_module.feed_forward", <class 'torchtitan.models.llama.model.FeedForward'>))                       
#     ('139761460312400', ("L['self']._checkpoint_wrapped_module.feed_forward.w2", <class 'float8_experimental.float8_dynamic_linear.Float8DynamicLinear'>))
triton_red_fused_abs_max_mul_silu_23 = async_compile.triton('triton_', '''                                                                                                                                                                                                            
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code
3. Adds a printout of the unique fwd and backward nn.Module stack
   elements, to make it easier to map the inductor kernel to the
   modeling code.  This is gated by `torch._inductor.config.comment_nn_module_stack_trace_info` and is off by default.

Example inductor kernel comment from LLaMa 3 8B training after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/km/ckm65ajvvhdoctx6djzhrkt7k5zmf3wyyugfqbhzyzplrd3eycxp.py                                                        
# Source Nodes: [silu, mul_6, abs_13, max_13], Original ATen: [aten.silu, aten.mul, aten.abs, aten.max]                                                     
# Source node to ATen node mapping:                                                                                                                         
#   abs_13 => abs_13                                                                                                                                        
#   max_13 => max_13                                                                                                                                        
#   mul_6 => mul_31                                                                                                                                         
#   silu => convert_element_type_58, convert_element_type_59, mul_26, sigmoid                                                                               
# Graph fragment:                                                                                                                                           
#     convert_element_type_58 = torch.ops.prims.convert_element_type.default(view_24, torch.float32)                                                        
#     sigmoid = torch.ops.aten.sigmoid.default(convert_element_type_58)                                                                                     
#     mul_26 = torch.ops.aten.mul.Tensor(convert_element_type_58, sigmoid);  convert_element_type_58 = sigmoid = None                                       
#     convert_element_type_59 = torch.ops.prims.convert_element_type.default(mul_26, torch.bfloat16);  mul_26 = None                                        
#     mul_31 = torch.ops.aten.mul.Tensor(convert_element_type_59, view_27);  convert_element_type_59 = view_27 = None                                       
#     abs_13 = torch.ops.aten.abs.default(mul_31)                                                                                                           
#     max_13 = torch.ops.aten.max.default(abs_13);  abs_13 = None                                                                                           
# Unique node to nn_module_stack last element                                                                                                               
#     ('139761509383760', ("L['self']._checkpoint_wrapped_module.feed_forward", <class 'torchtitan.models.llama.model.FeedForward'>))                       
#     ('139761460312400', ("L['self']._checkpoint_wrapped_module.feed_forward.w2", <class 'float8_experimental.float8_dynamic_linear.Float8DynamicLinear'>))
triton_red_fused_abs_max_mul_silu_23 = async_compile.triton('triton_', '''                                                                                                                                                                                                            
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code
3. Adds a printout of the unique fwd and backward nn.Module stack
   elements, to make it easier to map the inductor kernel to the
   modeling code.  This is gated by `torch._inductor.config.comment_nn_module_stack_trace_info` and is off by default.

Example inductor kernel comment from LLaMa 3 8B training after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/km/ckm65ajvvhdoctx6djzhrkt7k5zmf3wyyugfqbhzyzplrd3eycxp.py                                                        
# Source Nodes: [silu, mul_6, abs_13, max_13], Original ATen: [aten.silu, aten.mul, aten.abs, aten.max]                                                     
# Source node to ATen node mapping:                                                                                                                         
#   abs_13 => abs_13                                                                                                                                        
#   max_13 => max_13                                                                                                                                        
#   mul_6 => mul_31                                                                                                                                         
#   silu => convert_element_type_58, convert_element_type_59, mul_26, sigmoid                                                                               
# Graph fragment:                                                                                                                                           
#     convert_element_type_58 = torch.ops.prims.convert_element_type.default(view_24, torch.float32)                                                        
#     sigmoid = torch.ops.aten.sigmoid.default(convert_element_type_58)                                                                                     
#     mul_26 = torch.ops.aten.mul.Tensor(convert_element_type_58, sigmoid);  convert_element_type_58 = sigmoid = None                                       
#     convert_element_type_59 = torch.ops.prims.convert_element_type.default(mul_26, torch.bfloat16);  mul_26 = None                                        
#     mul_31 = torch.ops.aten.mul.Tensor(convert_element_type_59, view_27);  convert_element_type_59 = view_27 = None                                       
#     abs_13 = torch.ops.aten.abs.default(mul_31)                                                                                                           
#     max_13 = torch.ops.aten.max.default(abs_13);  abs_13 = None                                                                                           
# Unique node to nn_module_stack last element                                                                                                               
#     ('139761509383760', ("L['self']._checkpoint_wrapped_module.feed_forward", <class 'torchtitan.models.llama.model.FeedForward'>))                       
#     ('139761460312400', ("L['self']._checkpoint_wrapped_module.feed_forward.w2", <class 'float8_experimental.float8_dynamic_linear.Float8DynamicLinear'>))
triton_red_fused_abs_max_mul_silu_23 = async_compile.triton('triton_', '''                                                                                                                                                                                                            
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code
3. Adds a printout of the unique fwd and backward nn.Module stack
   elements, to make it easier to map the inductor kernel to the
   modeling code.  This is gated by `torch._inductor.config.comment_nn_module_stack_trace_info` and is off by default.

Example inductor kernel comment from LLaMa 3 8B training after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/km/ckm65ajvvhdoctx6djzhrkt7k5zmf3wyyugfqbhzyzplrd3eycxp.py                                                        
# Source Nodes: [silu, mul_6, abs_13, max_13], Original ATen: [aten.silu, aten.mul, aten.abs, aten.max]                                                     
# Source node to ATen node mapping:                                                                                                                         
#   abs_13 => abs_13                                                                                                                                        
#   max_13 => max_13                                                                                                                                        
#   mul_6 => mul_31                                                                                                                                         
#   silu => convert_element_type_58, convert_element_type_59, mul_26, sigmoid                                                                               
# Graph fragment:                                                                                                                                           
#     convert_element_type_58 = torch.ops.prims.convert_element_type.default(view_24, torch.float32)                                                        
#     sigmoid = torch.ops.aten.sigmoid.default(convert_element_type_58)                                                                                     
#     mul_26 = torch.ops.aten.mul.Tensor(convert_element_type_58, sigmoid);  convert_element_type_58 = sigmoid = None                                       
#     convert_element_type_59 = torch.ops.prims.convert_element_type.default(mul_26, torch.bfloat16);  mul_26 = None                                        
#     mul_31 = torch.ops.aten.mul.Tensor(convert_element_type_59, view_27);  convert_element_type_59 = view_27 = None                                       
#     abs_13 = torch.ops.aten.abs.default(mul_31)                                                                                                           
#     max_13 = torch.ops.aten.max.default(abs_13);  abs_13 = None                                                                                           
# Unique node to nn_module_stack last element                                                                                                               
#     ('139761509383760', ("L['self']._checkpoint_wrapped_module.feed_forward", <class 'torchtitan.models.llama.model.FeedForward'>))                       
#     ('139761460312400', ("L['self']._checkpoint_wrapped_module.feed_forward.w2", <class 'float8_experimental.float8_dynamic_linear.Float8DynamicLinear'>))
triton_red_fused_abs_max_mul_silu_23 = async_compile.triton('triton_', '''                                                                                                                                                                                                            
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo commented:
Ok, ready for review again, I found myself repeatedly coming back to this PR over the past weeks to help with inductor debugging and finally took the time to clean it up:
1. removed the nn.Module stack stuff for now as I haven't used that in months.  May come back to it in a future PR if the need arises.  Btw, the object IDs from this part of the logs were coming from activation checkpointing.
2. moved the graph fragment generation to use `n.format_node()`, it's a bit longer but removes the need to refactor FX code.  I think it would be actually really nice to generate code fragments, just don't want that to block this PR
3. made topological sort more efficient based on @ezyang 's suggestion

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code

Example float8-overhead-related inductor kernel comment after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/27/c27ts3rdw56ns7od5j6ovdnhxphished2lcu3adclzzixoo7khg5.py                                                                                      
# Source Nodes: [weight_fp8], Original ATen: [aten.mul, aten.clamp, aten._to_copy]                                                                                                        
# Source node to ATen node mapping:                                                                                                                                                       
#   weight_fp8 => clamp_max_1, clamp_min_3, convert_element_type_10, convert_element_type_11, convert_element_type_9, mul_3                                                               
# Graph fragment:                                                                                                                                                                         
#   %mul_3 : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%primals_2, %convert_element_type_8), kwargs = {})                                                   
#   %convert_element_type_9 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%mul_3, torch.float32), kwargs = {})                             
#   %clamp_min_3 : [num_users=1] = call_function[target=torch.ops.aten.clamp_min.default](args = (%convert_element_type_9, -448.0), kwargs = {})                                          
#   %clamp_max_1 : [num_users=1] = call_function[target=torch.ops.aten.clamp_max.default](args = (%clamp_min_3, 448.0), kwargs = {})                                                      
#   %convert_element_type_10 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%clamp_max_1, torch.bfloat16), kwargs = {})                     
#   %convert_element_type_11 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%convert_element_type_10, torch.float8_e4m3fn), kwargs = {})    
triton_poi_fused__to_copy_clamp_mul_5 = async_compile.triton('triton_', '''                                                                                                                                                                                                                                                                                                 
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code

Example float8-overhead-related inductor kernel comment after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/27/c27ts3rdw56ns7od5j6ovdnhxphished2lcu3adclzzixoo7khg5.py                                                                                      
# Source Nodes: [weight_fp8], Original ATen: [aten.mul, aten.clamp, aten._to_copy]                                                                                                        
# Source node to ATen node mapping:                                                                                                                                                       
#   weight_fp8 => clamp_max_1, clamp_min_3, convert_element_type_10, convert_element_type_11, convert_element_type_9, mul_3                                                               
# Graph fragment:                                                                                                                                                                         
#   %mul_3 : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%primals_2, %convert_element_type_8), kwargs = {})                                                   
#   %convert_element_type_9 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%mul_3, torch.float32), kwargs = {})                             
#   %clamp_min_3 : [num_users=1] = call_function[target=torch.ops.aten.clamp_min.default](args = (%convert_element_type_9, -448.0), kwargs = {})                                          
#   %clamp_max_1 : [num_users=1] = call_function[target=torch.ops.aten.clamp_max.default](args = (%clamp_min_3, 448.0), kwargs = {})                                                      
#   %convert_element_type_10 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%clamp_max_1, torch.bfloat16), kwargs = {})                     
#   %convert_element_type_11 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%convert_element_type_10, torch.float8_e4m3fn), kwargs = {})    
triton_poi_fused__to_copy_clamp_mul_5 = async_compile.triton('triton_', '''                                                                                                                                                                                                                                                                                                 
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code

Example float8-overhead-related inductor kernel comment after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/27/c27ts3rdw56ns7od5j6ovdnhxphished2lcu3adclzzixoo7khg5.py                                                                                      
# Source Nodes: [weight_fp8], Original ATen: [aten.mul, aten.clamp, aten._to_copy]                                                                                                        
# Source node to ATen node mapping:                                                                                                                                                       
#   weight_fp8 => clamp_max_1, clamp_min_3, convert_element_type_10, convert_element_type_11, convert_element_type_9, mul_3                                                               
# Graph fragment:                                                                                                                                                                         
#   %mul_3 : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%primals_2, %convert_element_type_8), kwargs = {})                                                   
#   %convert_element_type_9 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%mul_3, torch.float32), kwargs = {})                             
#   %clamp_min_3 : [num_users=1] = call_function[target=torch.ops.aten.clamp_min.default](args = (%convert_element_type_9, -448.0), kwargs = {})                                          
#   %clamp_max_1 : [num_users=1] = call_function[target=torch.ops.aten.clamp_max.default](args = (%clamp_min_3, 448.0), kwargs = {})                                                      
#   %convert_element_type_10 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%clamp_max_1, torch.bfloat16), kwargs = {})                     
#   %convert_element_type_11 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%convert_element_type_10, torch.float8_e4m3fn), kwargs = {})    
triton_poi_fused__to_copy_clamp_mul_5 = async_compile.triton('triton_', '''                                                                                                                                                                                                                                                                                                 
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ezyang commented on torch/_inductor/utils.py:
```diff
@@ -523,23 +523,53 @@ def get_kernel_metadata(node_schedule, wrapper):
+        detailed_metadata.append(f"{wrapper.comment} Graph fragment:")
+        for n in inductor_nodes:
+            # TODO(future): maybe refactor torch/fx/graph.py to make it easy to
+            # generate python code for graph fragments
```
yes please

ezyang reviewed (approved):
Thanks!

vkuzo committed:
Update on "Make inductor kernel metadata comments more descriptive"


Summary:

A couple of improvements to the generated comments in inductor kernels:

1. Makes the nodes in the comment topologically sorted, I think having them
   alphabetically sorted is a gotcha. I was always confused on why the
   sorting in the comments did not match the code.
2. Adds a printout of the aten graph fragment corresponding to the
   current inductor kernel, to make it easier to map from aten
   code to inductor code

Example float8-overhead-related inductor kernel comment after this PR:

```
# kernel path: /tmp/torchinductor_vasiliy/27/c27ts3rdw56ns7od5j6ovdnhxphished2lcu3adclzzixoo7khg5.py                                                                                      
# Source Nodes: [weight_fp8], Original ATen: [aten.mul, aten.clamp, aten._to_copy]                                                                                                        
# Source node to ATen node mapping:                                                                                                                                                       
#   weight_fp8 => clamp_max_1, clamp_min_3, convert_element_type_10, convert_element_type_11, convert_element_type_9, mul_3                                                               
# Graph fragment:                                                                                                                                                                         
#   %mul_3 : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%primals_2, %convert_element_type_8), kwargs = {})                                                   
#   %convert_element_type_9 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%mul_3, torch.float32), kwargs = {})                             
#   %clamp_min_3 : [num_users=1] = call_function[target=torch.ops.aten.clamp_min.default](args = (%convert_element_type_9, -448.0), kwargs = {})                                          
#   %clamp_max_1 : [num_users=1] = call_function[target=torch.ops.aten.clamp_max.default](args = (%clamp_min_3, 448.0), kwargs = {})                                                      
#   %convert_element_type_10 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%clamp_max_1, torch.bfloat16), kwargs = {})                     
#   %convert_element_type_11 : [num_users=1] = call_function[target=torch.ops.prims.convert_element_type.default](args = (%convert_element_type_10, torch.float8_e4m3fn), kwargs = {})    
triton_poi_fused__to_copy_clamp_mul_5 = async_compile.triton('triton_', '''                                                                                                                                                                                                                                                                                                 
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

vkuzo commented:
@pytorchbot merge

vkuzo commented:
@pytorchbot merge

jataylo commented:
@vkuzo @ezyang this commit broke rocm b2b_gemm tests cc: @jithunnair-amd @jeffdaily @pruthvistony @huydhn @malfet 
e.g. python test/inductor/test_b2b_gemm.py -k "test_b2b_gemm_right_assoc_good_shape"

We will need to revert this

this commit:
```
git log
commit 700a11fdd46f21a0c64c4ccbd24ac951e7b58307 (HEAD)
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/fx/node.py", line 70, in _find_module_of_method
    module = orig_method.__module__
AttributeError: 'functools.partial' object has no attribute '__module__'
```

this commit-1:
```
git log
commit 260e7cb143be4dc12e27931babcbcfaced3cfccc (HEAD, refs/bisect/good-260e7cb143be4dc12e27931babcbcfaced3cfccc)
.
----------------------------------------------------------------------
Ran 1 test in 21.012s

OK
```

I see this get_kernel_metadata being called:
`detailed_metadata.append(f"{wrapper.comment}   {n.format_node()}")`

Logs: [rocm-failure-b2b-gemm.txt](https://github.com/user-attachments/files/16678133/rocm-failure-b2b-gemm.txt)

malfet commented:
@jataylo can we simply skip this test on ROCM? Or are you saying that it simply doesn't work on Python-3.8? If that's the case, let's migrate to 3.9, as 3.8 will reach EOL in October this year

jataylo commented:
It's more than skipping a single test, there are multiple failures in the file regarding this. I hadn't associated this to being a 3.8 issue though, let me check if it works for us on 3.9/3.10 and see what our plan is there. cc: @jeffdaily @jithunnair-amd @pruthvistony

vkuzo commented:
I think a forward fix would be better than a revert, we can skip outputting this metadata in the cases where `n.format_node()` is not supported.  Is there a way to reproduce the breakage in CI?

jataylo commented:
> I think a forward fix would be better than a revert, we can skip outputting this metadata in the cases where `n.format_node()` is not supported. Is there a way to reproduce the breakage in CI?

Yep this one will fail with ciflow/rocm e.g.
```
https://hud.pytorch.org/failure?name=rocm%20%2F%20linux-focal-rocm6.1-py3.8%20%2F%20test%20(default%2C%205%2C%206%2C%20linux.rocm.gpu.2)&jobName=undefined&failureCaptures=%5B%22inductor%2Ftest_b2b_gemm.py%3A%3AB2BGEMMTest%3A%3Atest_b2b_gemm_trivial_right_assoc_good_shape%22%5D
```

A forward fix would be nice to have, we're also setting up a PR to  upgrade rocm CI to 3.10 which I'm hoping will resolve it too https://github.com/pytorch/pytorch/pull/134108
